### ISSUE TITLE
xUnit VS runner does not detect tests when AutoFac 4.0 is present

### ISSUE THREAD

**Issue by candychiu** [xunit/xunit#540]
I recently upgraded Autofac from 3.5 to 4.0.  After the upgrade, the VS runner can no longer detect tests.  I understand this may not be an xUnit issue.  In fact, if I restored Autofac to version 3.5, the latest prelease and stable version of xunit works.  With Autofac 4.0, both prelease and stable version doesn't work.  Are you guys aware of any potential collision?  

Environment: VS 2013 Update 5, .Net 4.5.1

**Comment by bradwilson** [comment]
This is the first I've heard of it. Do you have a repro project? I've never used Autofac. :)

**Comment by candychiu** [comment]
AutoFac is an IOC container: http://autofac.org/.  In fact someone wrote a connector between xUnit < 2.0 and Autofac.  You should be able to reproduce the issue by installing AutoFac from Nuget.

**Comment by bradwilson** [comment]
The newest release version of Autofac on NuGet is 3.5.2. There is no stable 4.0 version.

When attempting to install a pre-release version of Autofac in VS2015, NuGet refuses to install the package:

```
PM> install-package autofac -pre
Attempting to gather dependencies information for package 'autofac.4.0.0-beta6-110' with respect to project 'xunit540', targeting '.NETFramework,Version=v4.5.1'
Attempting to resolve dependencies for package 'autofac.4.0.0-beta6-110' with DependencyBehavior 'Lowest'
Resolving actions to install package 'autofac.4.0.0-beta6-110'
Resolved actions to install package 'autofac.4.0.0-beta6-110'
Adding package 'Autofac.4.0.0-beta6-110' to folder 'c:\Users\bradw\Documents\Visual Studio 2015\Projects\xunit540\packages'
Added package 'Autofac.4.0.0-beta6-110' to folder 'c:\Users\bradw\Documents\Visual Studio 2015\Projects\xunit540\packages'
Install failed. Rolling back...
Package 'Autofac 4.0.0-beta6-110' does not exist in project 'xunit540'
Removing package 'Autofac 4.0.0-beta6-110' from folder 'c:\Users\bradw\Documents\Visual Studio 2015\Projects\xunit540\packages'
Removed package 'Autofac 4.0.0-beta6-110' from folder 'c:\Users\bradw\Documents\Visual Studio 2015\Projects\xunit540\packages'
install-package : Failed to add reference to 'System.Collections'. Please make sure that it is in the Global Assembly Cache.
At line:1 char:1
+ install-package autofac -pre
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Install-Package], Exception
    + FullyQualifiedErrorId : NuGetCmdletUnhandledException,NuGet.PackageManagement.PowerShellCmdlets.InstallPackageCommand
```

I will need to make a VM with VS2013 to see if I can reproduce the issue there, as I don't have VS2013 on my PC.

**Comment by bradwilson** [comment]
I had better success with VS2013 being able to install the pre-release version of Autofac. However, I believe that I understand the problem you may be having (though I am getting a slightly different problem than you've described).

Here is the only test in the project:

```csharp
using Autofac;
using Xunit;

namespace xunit540
{
    public class Class1
    {
        [Fact]
        public void TheTest()
        {
            new ContainerBuilder();
        }
    }
}
```

I have installed `xunit 2.0.0` and `xunit.runner.visualstudio 2.1.0`. With `xutofac 3.5.2` installed, this test is discovered and runs fine. With `Autofac 4.0.0-beta6-110` installed, I was able to get the test compiled and discovered, but it failed to run with the following error:

```
Test Name:	xunit540.Class1.TheTest
Test FullName:	xunit540.Class1.TheTest
Test Source:	c:\Users\Brad\Documents\Visual Studio 2013\Projects\xunit540\xunit540\Class1.cs : line 15
Test Outcome:	Failed
Test Duration:	0:00:00.008

Result Message:	System.IO.FileLoadException : Could not load file or assembly 'Autofac, Version=4.0.0.0, Culture=neutral, PublicKeyToken=8f68bdc6b69987bd' or one of its dependencies. Strong name signature could not be verified.  The assembly may have been tampered with, or it was delay signed but not fully signed with the correct private key. (Exception from HRESULT: 0x80131045)
Result StackTrace:	at xunit540.Class1.TheTest()
```

Given that the Autofac assembly does not appear to have a valid signature, it would seem that all bets are off on what works and what doesn't. It wouldn't surprise me to discover that you were having test discovery issues (you probably use Autofac more than I did in my sample), if there was an assembly loading error during discovery.

At this point, I feel comfortable saying that this is very likely an external issue (with Autofac) rather than an xUnit.net issue.

**Comment by candychiu** [comment]
Thanks for investigating.

**Comment by tillig** [comment]
For reference, this appears to be resolved with the Autofac 4.0.0-beta7-130 package.
